File: stock-advisor/src/advisor/news.py
"""News ingestion and event classification (mirrors live-console/news.mjs).

Headlines are third-party claims, not verified fact. This module classifies and
weights them so downstream logic can *reason* about them; it never treats a
headline as a figure.
"""
from __future__ import annotations

import re
import urllib.parse
from dataclasses import dataclass, asdict
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime

import requests

FEED = ("https://news.google.com/rss/search?q={q}&hl=en-IN&gl=IN&ceid=IN:en")
TIMEOUT = 20

# (key, weight, label, pattern) — order matters: specific before generic.
EVENTS = [
    ("fraud", -3, "Fraud or investigation",
     r"\b(fraud|scam|siphon|forensic audit|sebi bars|insider trading|money launder)"),
    ("default", -3, "Credit or solvency event",
     r"\b(default|insolven|nclt|bankrupt|debt restructur|rating cut|downgraded to)"),
    ("regulator", -2, "Regulatory action",
     r"\b(sebi (notice|order|penalt)|rbi (action|penalt|restrict)|"
     r"usfda (warning|import alert|observation)|show cause|penalt)"),
    ("litigation", -2, "Litigation or tax demand",
     r"\b(lawsuit|litigation|court (order|case)|tribunal|arbitration|tax demand|gst notice)"),
    ("pledge", -2, "Promoter selling or pledging",
     r"\b(pledg|promoter (stake sale|sells|selling|offload))"),
    ("downgrade", -2, "Analyst downgrade",
     r"\b(downgrade|cuts? target|slash(es)? target|underperform|sell rating)"),
    ("profitfall", -2, "Weak results",
     r"\b(profit (falls|drops|declines|slumps|plunges)|loss widens|posts loss|"
     r"misses estimates|weak (results|quarter))"),
    ("exit", -2, "Senior management exit",
     r"\b(resigns|steps down|quits|ceo exit|cfo exit|auditor resign)"),
    ("upgrade", 2, "Analyst upgrade",
     r"\b(upgrade|raises? target|hikes? target|outperform|buy rating|top pick)"),
    ("profitrise", 2, "Strong results",
     r"\b(profit (rises|jumps|surges|soars)|beats estimates|strong (results|quarter)|"
     r"record (profit|revenue))"),
    ("order", 2, "Order or contract win",
     r"\b(order win|order book|(bags?|wins?|secures?|receives?)\b[^.]{0,45}?"
     r"\b(order|contract|deal|mandate|tender))"),
    ("expansion", 1, "Expansion or M&A",
     r"\b(capex|new plant|capacity expansion|acquisition|acquires|merger|joint venture)"),
    ("policy", 1, "Policy or government action",
     r"\b(government|policy|pli|subsid|budget|duty|tariff|gst rate|repo rate)"),
    ("payout", 1, "Shareholder payout", r"\b(dividend|buyback|bonus issue|stock split)"),
    ("results", 0, "Results scheduled or reported",
     r"\b(q[1-4] results|quarterly results|earnings|board meeting)"),
]
COMPILED = [(k, w, l, re.compile(p, re.I)) for k, w, l, p in EVENTS]


@dataclass
class NewsItem:
    title: str
    link: str
    source: str
    published: str | None
    age_days: float | None
    event: str
    event_label: str
    weight: int

    def to_dict(self) -> dict:
        return asdict(self)


def classify(title: str) -> tuple[str, int, str]:
    for key, weight, label, pattern in COMPILED:
        if pattern.search(title):
            return key, weight, label
    return "general", 0, "General coverage"


def _strip(x: str) -> str:
    return re.sub(r"<[^>]+>", " ", x).replace("&amp;", "&").replace("&#39;", "'").strip()


def fetch_news(symbol: str, name: str | None = None, days: int = 14,
               limit: int = 20) -> list[NewsItem]:
    """Company news. Returns [] when the feed is unreachable — never invents items."""
    query = f'"{name or symbol}" (share OR stock OR results OR NSE) when:{days}d'
    url = FEED.format(q=urllib.parse.quote(query))
    resp = requests.get(url, timeout=TIMEOUT,
                        headers={"User-Agent": "advisor-360/1.0",
                                 "Accept": "application/rss+xml, application/xml"})
    resp.raise_for_status()
    items: list[NewsItem] = []
    now = datetime.now(timezone.utc)
    for block in re.findall(r"<item>(.*?)</item>", resp.text, re.S)[:limit]:
        title = _strip(re.search(r"<title>(.*?)</title>", block, re.S).group(1)) \
            if re.search(r"<title>(.*?)</title>", block, re.S) else ""
        if not title:
            continue
        link_m = re.search(r"<link>(.*?)</link>", block, re.S)
        pub_m = re.search(r"<pubDate>(.*?)</pubDate>", block, re.S)
        src_m = re.search(r"<source[^>]*>(.*?)</source>", block, re.S)
        age = None
        published = _strip(pub_m.group(1)) if pub_m else None
        if published:
            try:
                age = round((now - parsedate_to_datetime(published)).total_seconds() / 86400, 1)
            except (TypeError, ValueError):
                age = None
        key, weight, label = classify(title)
        items.append(NewsItem(title=title, link=_strip(link_m.group(1)) if link_m else "",
                              source=_strip(src_m.group(1)) if src_m else "news",
                              published=published, age_days=age,
                              event=key, event_label=label, weight=weight))
    return items


def pressure(items: list[NewsItem]) -> dict:
    """Recency-weighted net direction of the news flow."""
    net = 0.0
    intensity = 0.0
    for i in items:
        if not i.weight:
            continue
        decay = 0.5 if i.age_days is None else max(0.15, 1 - i.age_days / 14)
        net += i.weight * decay
        intensity += abs(i.weight) * decay
    tone = "positive" if net > 1.5 else "negative" if net < -1.5 else "mixed"
    return {"net": round(net, 2), "intensity": round(intensity, 2), "tone": tone,
            "material_count": sum(1 for i in items if abs(i.weight) >= 2)}

File: live-console/netlify/functions/news.mjs
// GET /api/news?symbol=RELIANCE        — company news
// GET /api/news?market=1               — market-wide news
// Source: Google News RSS (India edition) — reliable from cloud IPs, unlike the
// exchange APIs. Headlines are classified into event types and scored for
// direction, so the verdict engine can reason about them rather than just list them.

import { jsonResponse, stripTags } from "./_lib.mjs";

const FEEDS = {
  market: "https://news.google.com/rss/search?q=" +
    encodeURIComponent("Nifty OR Sensex stock market India when:3d") +
    "&hl=en-IN&gl=IN&ceid=IN:en",
};

function companyFeed(symbol, name) {
  const q = `"${name || symbol}" (share OR stock OR results OR NSE) when:14d`;
  return "https://news.google.com/rss/search?q=" + encodeURIComponent(q) +
    "&hl=en-IN&gl=IN&ceid=IN:en";
}

/* Event taxonomy. Weight is the direction and force of the event, -3..+3.
   Order matters: the first pattern that matches wins, so specific beats generic. */
const EVENTS = [
  { k: "fraud",        w: -3, re: /\b(fraud|scam|siphon|forensic audit|sebi bars|insider trading|money launder)/i,
    label: "Fraud or investigation" },
  { k: "default",      w: -3, re: /\b(default|insolven|nclt|bankrupt|debt restructur|credit rating cut|downgrade[ds]? to)/i,
    label: "Credit or solvency event" },
  { k: "regulator",    w: -2, re: /\b(sebi (?:notice|order|penalt)|rbi (?:action|penalt|restrict)|usfda (?:warning|import alert|observation)|show cause|penalt)/i,
    label: "Regulatory action" },
  { k: "litigation",   w: -2, re: /\b(lawsuit|litigation|court (?:order|case)|tribunal|arbitration|tax demand|gst notice)/i,
    label: "Litigation or tax demand" },
  { k: "pledge",       w: -2, re: /\b(pledg|promoter (?:stake sale|sells|selling|offload))/i,
    label: "Promoter selling or pledging" },
  { k: "downgrade",    w: -2, re: /\b(downgrade|cuts? target|slash(?:es)? target|underperform|reduce rating|sell rating)/i,
    label: "Analyst downgrade" },
  { k: "profitfall",   w: -2, re: /\b(profit (?:falls|drops|declines|slumps|plunges)|loss widens|posts loss|misses estimates|weak (?:results|quarter))/i,
    label: "Weak results" },
  { k: "exit",         w: -2, re: /\b(resigns|steps down|quits|ceo exit|cfo exit|auditor resign)/i,
    label: "Senior management exit" },
  { k: "upgrade",      w: 2,  re: /\b(upgrade|raises? target|hikes? target|outperform|buy rating|top pick)/i,
    label: "Analyst upgrade" },
  { k: "profitrise",   w: 2,  re: /\b(profit (?:rises|jumps|surges|soars|up \d)|beats estimates|strong (?:results|quarter)|record (?:profit|revenue))/i,
    label: "Strong results" },
  { k: "order",        w: 2,  re: /\b(order win|order book|(?:bags?|wins?|secures?|receives?)\b[^.]{0,45}?\b(?:order|contract|deal|mandate|tender))/i,
    label: "Order or contract win" },
  { k: "expansion",    w: 1,  re: /\b(capex|new plant|capacity expansion|acquisition|acquires|merger|joint venture|stake buy)/i,
    label: "Expansion or M&A" },
  { k: "policy",       w: 1,  re: /\b(government|policy|pli|subsid|budget|duty|tariff|gst rate|rbi (?:policy|rate)|repo rate)/i,
    label: "Policy or government action" },
  { k: "payout",       w: 1,  re: /\b(dividend|buyback|bonus issue|stock split)/i,
    label: "Shareholder payout" },
  { k: "results",      w: 0,  re: /\b(q[1-4] results|quarterly results|earnings|board meeting)/i,
    label: "Results scheduled or reported" },
];

function classify(title) {
  for (const e of EVENTS) if (e.re.test(title)) return e;
  return { k: "general", w: 0, label: "General coverage" };
}

function parseRss(xml, limit) {
  const items = [];
  for (const m of xml.matchAll(/<item>([\s\S]*?)<\/item>/g)) {
    const block = m[1];
    const title = stripTags((block.match(/<title>([\s\S]*?)<\/title>/) || [, ""])[1]);
    if (!title) continue;
    const link = stripTags((block.match(/<link>([\s\S]*?)<\/link>/) || [, ""])[1]);
    const pub = stripTags((block.match(/<pubDate>([\s\S]*?)<\/pubDate>/) || [, ""])[1]);
    const src = stripTags((block.match(/<source[^>]*>([\s\S]*?)<\/source>/) || [, ""])[1]);
    const ev = classify(title);
    const ageDays = pub ? (Date.now() - new Date(pub).getTime()) / 86400000 : null;
    items.push({
      title, link, published: pub, source: src || "news",
      ageDays: ageDays == null ? null : +ageDays.toFixed(1),
      event: ev.k, eventLabel: ev.label, weight: ev.w,
    });
    if (items.length >= (limit || 25)) break;
  }
  return items;
}

/** Net news pressure: recent and strong events count for more. */
function pressure(items) {
  let score = 0, weighted = 0;
  items.forEach((i) => {
    if (!i.weight) return;
    const decay = i.ageDays == null ? 0.5 : Math.max(0.15, 1 - i.ageDays / 14);
    score += i.weight * decay;
    weighted += Math.abs(i.weight) * decay;
  });
  return {
    net: +score.toFixed(2),
    intensity: +weighted.toFixed(2),
    tone: score > 1.5 ? "positive" : score < -1.5 ? "negative" : "mixed",
  };
}

export default async (req) => {
  const url = new URL(req.url);
  const symbol = (url.searchParams.get("symbol") || "").toUpperCase().trim();
  const name = url.searchParams.get("name") || "";
  const market = url.searchParams.get("market");

  if (!market && !/^[A-Z0-9&.-]{1,20}$/.test(symbol)) {
    return jsonResponse({ error: "bad symbol" }, { status: 400, maxAge: 60 });
  }
  const feed = market ? FEEDS.market : companyFeed(symbol, name);
  try {
    const r = await fetch(feed, {
      headers: { "User-Agent": "Mozilla/5.0 (advisor-360 personal console)",
                 "Accept": "application/rss+xml, application/xml, text/xml" },
    });
    if (!r.ok) throw new Error(`feed ${r.status}`);
    const items = parseRss(await r.text(), market ? 18 : 20);
    return jsonResponse({
      symbol: market ? "MARKET" : symbol,
      items,
      pressure: pressure(items),
      materialCount: items.filter((i) => Math.abs(i.weight) >= 2).length,
    }, { maxAge: 900 });
  } catch (e) {
    return jsonResponse({ symbol: symbol || "MARKET", error: "news unavailable",
                          detail: e.message, items: [] }, { maxAge: 120 });
  }
};

export const config = { path: "/api/news" };
